nanofi/src/core/string_utils.c

```diff
@@ -0,0 +1,87 @@
+/**
+ *
+ * Licensed to the Apache Software Foundation (ASF) under one or more
+ * contributor license agreements.  See the NOTICE file distributed with
+ * this work for additional information regarding copyright ownership.
+ * The ASF licenses this file to You under the Apache License, Version 2.0
+ * (the "License"); you may not use this file except in compliance with
+ * the License.  You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "core/cstructs.h"
+#include "core/string_utils.h"
+#include <string.h>
+#include <stdlib.h>
+
+tokens * tokenize_string(char * str, const char * delimiter) {
+    if (!str || !delimiter) return NULL;
+
+    char delim = delimiter[0];
+    if (delim == '\\') {
+          if (strlen(delimiter) > 1) {
+            switch (delimiter[1]) {
+              case 'r':
+                delim = '\r';
+                break;
+              case 't':
+                delim = '\t';
+                break;
+              case 'n':
+                delim = '\n';
+                break;
+              case '\\':
+                delim = '\\';
+                break;
+              default:
+                break;
+            }
+        }
+    }
+
+    char * begin = str;
+    char * end = NULL;
+    int num_strings = 0;
+    while ((end = strchr(begin, delim))) {
+        if (begin == end) {
+            begin++;
+            continue;
+        }
+        begin = (end+1);
+        num_strings++;
+    }
+
+    struct tokens * tks = (struct tokens *)malloc(sizeof(tokens));
+    tks->str_list = malloc(sizeof(char *) * (num_strings));
+    tks->num_strings = 0;
+    tks->total_bytes = 0;
+
+    for (int i = 0; i < num_strings; ++i) {
```

Comment:
If you allocate the block with calloc, you get it filled with zeros, so you don't need this at all.